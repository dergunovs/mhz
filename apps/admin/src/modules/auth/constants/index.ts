export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
