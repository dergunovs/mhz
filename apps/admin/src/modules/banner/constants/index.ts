export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
