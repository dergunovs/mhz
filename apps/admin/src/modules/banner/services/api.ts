import { ComputedRef, Ref } from 'vue';

import { IPageQuery, api, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export async function getBannersApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

  return data;
}

export async function getBannerApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

  return data.data;
}

export async function postBannerApi(formData: IBanner) {
  const { data } = await api.post<IBaseReply>(API_BANNER, formData);

  return data;
}

export async function updateBannerApi(formData: IBanner) {
  const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

  return data;
}

export async function deleteBannerApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

  return data;
}
