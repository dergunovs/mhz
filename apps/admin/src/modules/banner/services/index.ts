import { ComputedRef, Ref } from 'vue';

import { api, useQuery, useMutation, IPageQuery, convertParams } from 'mhz-helpers';
import { API_BANNER, IBanner, IBaseReply, IFilterData } from 'mhz-contracts';

export function getBanners(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: IBanner[]; total: number; filters: IFilterData }>(API_BANNER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_BANNER, query], queryFn: fn });
}

export function getBanner(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: IBanner }>(`${API_BANNER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_BANNER, id], queryFn: fn });
}

export function postBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.post<IBaseReply>(API_BANNER, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function updateBanner(options: object) {
  async function fn(formData: IBanner) {
    const { data } = await api.patch<IBaseReply>(`${API_BANNER}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}

export function deleteBanner(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_BANNER}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_BANNER], mutationFn: fn, ...options });
}
