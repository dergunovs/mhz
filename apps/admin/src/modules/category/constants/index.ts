export const URL_CATEGORY = '/categories';
export const URL_CATEGORY_CREATE = `${URL_CATEGORY}/create`;
export const URL_CATEGORY_EDIT = `${URL_CATEGORY}/edit`;

export const CATEGORY_FIELD_TYPE_OPTIONS = ['string', 'number', 'boolean'];
