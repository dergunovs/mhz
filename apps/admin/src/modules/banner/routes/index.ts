import { URL_BANNER, URL_BANNER_CREATE, URL_BANNER_EDIT } from '@/banner/constants';

export const bannerRoutes = [
  {
    path: URL_BANNER,
    name: 'BannerList',
    component: () => import('@/banner/pages/BannerListPage.vue'),
  },
  {
    path: URL_BANNER_CREATE,
    name: 'BannerCreate',
    component: () => import('@/banner/pages/BannerCreatePage.vue'),
  },
  {
    path: `${URL_BANNER_EDIT}/:banner`,
    name: 'BannerEdit',
    component: () => import('@/banner/pages/BannerEditPage.vue'),
  },
];
