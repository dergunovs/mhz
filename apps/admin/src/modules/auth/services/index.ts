import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
