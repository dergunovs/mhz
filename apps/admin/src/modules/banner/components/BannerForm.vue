<template>
  <form @submit.prevent="props.banner?._id ? update() : submit()" :class="$style.form">
    <UiField label="Text" isRequired :error="error('text')">
      <UiInput v-model="formData.text" isFocus />
    </UiField>

    <UiField label="Color" isRequired :error="error('color')">
      <UiInput v-model="formData.color" />
    </UiField>

    <UiField label="Product" isRequired :error="error('product')">
      <UiSelect
        v-model="formData.product"
        :options="allProducts"
        @reachedBottom="scrollProducts(isLoadingProducts, setProductsPage(productsPage + 1, productsPage))"
      />
    </UiField>

    <UiCheckbox label="Is active" v-model="formData.isActive" isRequired :error="error('isActive')" />

    <UiUpload
      label="Upload image"
      :file="imageFile"
      isSingle
      isRequired
      :isDisabled="!!formData.imageUrl"
      :extensions="['png']"
      :error="error('imageUrl')"
      @add="addImageFile"
      @remove="removeImageFile"
      @upload="mutateUploadFile(imageFile)"
    />

    <ImagePreview v-if="formData.imageUrl" :urls="[formData.imageUrl]" @delete="formData.imageUrl = ''" />

    <FormButtons :id="props.banner?._id" :isLoading="isLoadingPost || isLoadingUpdate" @delete="handleDelete" />
  </form>
</template>

<script setup lang="ts">
import { ref, computed, onMounted, watch } from 'vue';
import { useRouter } from 'vue-router';

import { UiField, UiInput, UiUpload, UiSelect, UiCheckbox, toast } from 'mhz-ui';
import { clone, useValidator, required, useQueryClient, useInfiniteScroll, usePagination } from 'mhz-helpers';
import { API_BANNER, IBanner, IProduct } from 'mhz-contracts';

import ImagePreview from '@/common/components/ImagePreview.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { URL_BANNER } from '@/banner/constants';
import { postBanner, updateBanner, deleteBanner } from '@/banner/services';
import { uploadFile } from '@/common/services';
import { getProducts } from '@/product/services';

interface IProps {
  banner?: IBanner;
}

const props = defineProps<IProps>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<IBanner>({
  text: '',
  isActive: true,
  imageUrl: '',
  color: '',
  product: {} as IProduct,
});

const { mutate: mutatePost, isPending: isLoadingPost } = postBanner({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_BANNER] });
    toast.success('Banner added');
    router.push(URL_BANNER);
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = updateBanner({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_BANNER] });
    toast.success('Banner updated');
  },
});

const { mutate: mutateDelete } = deleteBanner({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_BANNER] });
    await queryClient.refetchQueries({ queryKey: [API_BANNER] });
    toast.success('Banner deleted');
    router.push(URL_BANNER);
  },
});

const rules = computed(() => {
  return {
    text: required,
    isActive: required,
    imageUrl: required,
    color: required,
  };
});

const { error, isValid } = useValidator(formData, rules);

function submit() {
  if (isValid()) mutatePost(formData.value);
}

function update() {
  if (isValid()) mutateUpdate(formData.value);
}

function handleDelete(id: string) {
  mutateDelete(id);
}

const imageFile = ref<File>();

function addImageFile(file: File) {
  imageFile.value = file;
}

function removeImageFile() {
  imageFile.value = undefined;
}

const { mutate: mutateUploadFile } = uploadFile({
  onSuccess: (data: string) => {
    formData.value.imageUrl = data;
    removeImageFile();
    toast.success('Icon added');
  },
});

const {
  page: productsPage,
  allData: allProducts,
  addData: addProducts,
  handleScroll: scrollProducts,
} = useInfiniteScroll<IProduct>();

const { data: productsData, isLoading: isLoadingProducts } = getProducts(productsPage);
const { data: products, setPage: setProductsPage } = usePagination(productsData);

watch(
  () => products.value,
  () => {
    if (products.value && !isLoadingProducts.value) addProducts(products.value);
  }
);

onMounted(() => {
  if (props.banner) formData.value = clone(props.banner);
  if (products.value) addProducts(products.value);
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: flex-start;
}
</style>
