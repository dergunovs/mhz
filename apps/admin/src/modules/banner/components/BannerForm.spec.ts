import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { API_BANNER, IBanner, IBaseReply } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';

import BannerForm from './BannerForm.vue';

import { wrapperFactory, mockQueryReply, mockMutationReply, router } from '@/common/test';
import { BANNERS } from '@/banner/fixtures';
import { PRODUCTS } from '@/product/fixtures';
import { URL_BANNER } from '@/banner/constants';
import * as productServices from '@/product/services';
import * as bannerServices from '@/banner/services';

const spyGetProducts = vi.spyOn(productServices, 'getProducts').mockReturnValue(mockQueryReply(PRODUCTS));

let onSuccessPost: () => void;
const spyMutatePost = vi.fn();

vi.spyOn(bannerServices, 'postBanner').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessPost = options.onSuccess;

  return mockMutationReply<IBaseReply, IBanner>(spyMutatePost);
});

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(bannerServices, 'updateBanner').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessUpdate = options.onSuccess;

  return mockMutationReply<IBaseReply, IBanner>(spyMutateUpdate);
});

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(bannerServices, 'deleteBanner').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyToastSuccess = vi.spyOn(toast, 'success');

const BANNER = BANNERS.data[0];
const QUERY_KEY = { queryKey: [API_BANNER] };

let wrapper: VueWrapper;

const bannerForm = '[data-test="banner-form"]';
const bannerFormText = '[data-test="banner-form-text"]';
const bannerFormColor = '[data-test="banner-form-color"]';
const bannerFormProduct = '[data-test="banner-form-product"]';
const bannerFormIsActive = '[data-test="banner-form-is-active"]';
const bannerFormImagePreview = '[data-test="banner-form-image-preview"]';
const bannerFormButtons = '[data-test="banner-form-buttons"]';

beforeEach(() => {
  wrapper = wrapperFactory(BannerForm, {
    props: {
      banner: BANNER,
    },
  });
});

enableAutoUnmount(afterEach);

describe('BannerForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(BannerForm)).toBeTruthy();
  });

  it('shows banner content in fields if props is passed', async () => {
    expect(wrapper.findComponent(bannerFormText).attributes('modelvalue')).toBe(BANNER.text);
    expect(wrapper.findComponent(bannerFormColor).attributes('modelvalue')).toBe(BANNER.color);
    expect(wrapper.findComponent(bannerFormProduct).attributes('data-product-id')).toBe(BANNER.product._id);
    expect(wrapper.findComponent(bannerFormIsActive).attributes('modelvalue')).toBe(BANNER.isActive.toString());
    expect(wrapper.findComponent(bannerFormImagePreview).attributes('urls')).toBe(BANNER.imageUrl);
  });

  it('shows image preview if banner has image', async () => {
    expect(wrapper.findComponent(bannerFormImagePreview).exists()).toBe(true);

    await wrapper.setProps({ banner: undefined });

    expect(wrapper.findComponent(bannerFormImagePreview).exists()).toBe(true);
  });

  it('gets products to fill the select options', async () => {
    expect(spyGetProducts).toBeCalledTimes(1);
  });

  it('updates banner', async () => {
    const NEW_BANNER_TEXT = 'yo';

    await wrapper.findComponent(bannerFormText).setValue(NEW_BANNER_TEXT);

    await wrapper.find(bannerForm).trigger('submit');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ ...BANNER, text: NEW_BANNER_TEXT });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes banner', async () => {
    wrapper.findComponent<DefineComponent>(bannerFormButtons).vm.$emit('delete', BANNER._id);

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(BANNER._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_BANNER);
  });

  it('creates banner', async () => {
    wrapper.unmount();

    const wrapperWithoutBanner = wrapperFactory(BannerForm, {});

    const NEW_BANNER = {
      color: '#000',
      imageUrl: 'pic.jpg',
      isActive: false,
      product: PRODUCTS.data[0],
      text: 'new text',
    };

    await wrapperWithoutBanner.findComponent(bannerFormColor).setValue(NEW_BANNER.color);
    await wrapperWithoutBanner.findComponent(bannerFormIsActive).setValue(NEW_BANNER.isActive);
    await wrapperWithoutBanner.findComponent(bannerFormText).setValue(NEW_BANNER.text);

    wrapperWithoutBanner.vm.formData.imageUrl = NEW_BANNER.imageUrl;
    wrapperWithoutBanner.vm.formData.product = NEW_BANNER.product;

    await wrapperWithoutBanner.find(bannerForm).trigger('submit');

    expect(spyMutatePost).toBeCalledTimes(1);
    expect(spyMutatePost).toBeCalledWith(NEW_BANNER);

    onSuccessPost();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_BANNER);
  });
});
