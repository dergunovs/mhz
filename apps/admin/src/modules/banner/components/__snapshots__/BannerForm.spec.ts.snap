// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BannerForm > matches snapshot 1`] = `
"<form class="_form_6d793d" data-test="banner-form">
  <fieldset label="Text" isrequired=""><input type="text" modelvalue="Too fast! It will exceed your expectations!" isfocus="" data-test="banner-form-text"></fieldset>
  <fieldset label="Color" isrequired=""><input type="text" modelvalue="#FFEFDB" data-test="banner-form-color"></fieldset>
  <fieldset label="Product" isrequired=""><input type="text" modelvalue="[object Object]" options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" lang="en" data-test="banner-form-product" data-product-id="64de340eb9d96f9e20bef523"></fieldset><input type="checkbox" label="Is active" modelvalue="true" isrequired="" data-test="banner-form-is-active"><input type="file" label="Upload image" issingle="" isrequired="" isdisabled="true" extensions="png" lang="en">
  <image-preview-stub urls="1696053502949-980.png" isthumb="false" data-test="banner-form-image-preview"></image-preview-stub>
  <form-buttons-stub id="6517b90000009354ff5e5a2a" data-test="banner-form-buttons"></form-buttons-stub>
</form>"
`;
